pipeline/run_pipeline.py

```diff
@@ -39,186 +93,533 @@ def load_and_sample_datasets(cfg):
+            model_base.tokenize_instructions_fn,
+            model_base.refusal_toks,
+        )
+        harmless_val_scores = get_refusal_scores(
+            model_base.model,
+            harmless_val,
+            model_base.tokenize_instructions_fn,
+            model_base.refusal_toks,
+        )
+        harmful_val = filter_examples(
+            harmful_val, harmful_val_scores, 0, lambda x, y: x > y
+        )
+        harmless_val = filter_examples(
+            harmless_val, harmless_val_scores, 0, lambda x, y: x < y
+        )
+
     return harmful_train, harmless_train, harmful_val, harmless_val
 
-def generate_and_save_candidate_directions(cfg, model_base, harmful_train, harmless_train):
+
+def generate_and_save_candidate_directions(
+    cfg, model_base, harmful_train, harmless_train
+):
     """Generate and save candidate directions."""
-    if not os.path.exists(os.path.join(cfg.artifact_path(), 'generate_directions')):
-        os.makedirs(os.path.join(cfg.artifact_path(), 'generate_directions'))
+    if not os.path.exists(os.path.join(cfg.artifact_path(), "generate_directions")):
+        os.makedirs(os.path.join(cfg.artifact_path(), "generate_directions"))
 
     mean_diffs = generate_directions(
         model_base,
         harmful_train,
         harmless_train,
-        artifact_dir=os.path.join(cfg.artifact_path(), "generate_directions"))
+        artifact_dir=os.path.join(cfg.artifact_path(), "generate_directions"),
+    )
 
-    torch.save(mean_diffs, os.path.join(cfg.artifact_path(), 'generate_directions/mean_diffs.pt'))
+    torch.save(
+        mean_diffs,
+        os.path.join(cfg.artifact_path(), "generate_directions/mean_diffs.pt"),
+    )
 
     return mean_diffs
 
-def select_and_save_direction(cfg, model_base, harmful_val, harmless_val, candidate_directions):
+
+def normalize_vector(vector):
+    """Normalize a given vector to have unit length."""
+    norm = torch.norm(vector)
+    if norm == 0:
+        return vector
+    return vector / norm
+
+
+def plotting_refusal_with_activations(
+    layers, average_dot_products, resultant_dot_products, activations_dir
+):
+    layers=range(layers)
+    plt.figure(figsize=(10, 5))  # Set the figure size as needed
+
+    # Plotting Average Dot Products
+    plt.plot(layers, average_dot_products, label="Average Dot Product", marker="o")
+
+    # Plotting Resultant Dot Products
+    plt.plot(layers, resultant_dot_products, label="Resultant Dot Product", marker="o")
+
+    plt.title("Dot Products Comparison Across Layers")  # Title of the plot
+    plt.xlabel("Layer Number")  # Label for the x-axis
+    plt.ylabel("Dot Product Value")  # Label for the y-axis
+    plt.xticks(layers)  # Ensure all layer numbers are marked
+    plt.legend()
+
+    plot_file_path = os.path.join(activations_dir, "dot_products_plot.png")
+    plt.savefig(plot_file_path)
+    plt.close()
+    print(f"Plot saved to {plot_file_path}")
+
+
+def computing_dot_product(
+    data_list,
```

Comment:
It would be good to have types next to the arguments

Reply:
+ a better name. `data_list`  doesn't convey anything. `layer_wise_activations` is a better name for example.